File: ts/test/cross-language.test.ts
/**
 * Cross-language integration tests
 * Tests TypeScript clients against Go servers
 */

import { test, expect, describe, beforeAll, afterAll } from "bun:test";
import { Session } from "../src/session.ts";
import { HttpClient } from "../src/http-client.ts";
import { WsClient } from "../src/ws-client.ts";
import { Peer } from "../src/peer.ts";
import { MethodMap } from "../src/helpers.ts";
import { newRequest } from "../src/types.ts";

// Start the Go test server
let goServerProcess: ReturnType<typeof Bun.spawn> | null = null;

const HTTP_PORT = 18080;
const WS_PORT = 18081;
const PEER_PORT = 18082;

// Helper to wait for a port to be ready
async function waitForPort(port: number, maxAttempts = 20, delayMs = 100): Promise<void> {
  for (let i = 0; i < maxAttempts; i++) {
    try {
      const conn = await Bun.connect({
        hostname: "localhost",
        port,
        socket: {
          data() {},
          open(socket) {
            socket.end();
          },
        },
      });
      conn.end();
      return;
    } catch (err) {
      if (i === maxAttempts - 1) {
        throw new Error(`Port ${port} not ready after ${maxAttempts} attempts`);
      }
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
  }
}

beforeAll(async () => {
  // Start Go server
  const goServerPath = "../go/jsonrpc3/testserver";

  goServerProcess = Bun.spawn(
    [goServerPath, "-http", HTTP_PORT.toString(), "-ws", WS_PORT.toString(), "-peer", PEER_PORT.toString()],
    {
      stdout: "pipe",
      stderr: "pipe",
    }
  );

  // Wait for all server ports to be ready
  await waitForPort(HTTP_PORT);
  await waitForPort(WS_PORT);
  await waitForPort(PEER_PORT);

  console.log("Go test server started on ports HTTP:", HTTP_PORT, "WS:", WS_PORT, "Peer:", PEER_PORT);
});

afterAll(() => {
  if (goServerProcess) {
    goServerProcess.kill();
    console.log("Go test server stopped");
  }
});

describe("Cross-Language HTTP", () => {
  test("TypeScript HTTP client can call Go HTTP server - basic add", async () => {
    const session = new Session();
    const client = new HttpClient(`http://localhost:${HTTP_PORT}`, session);

    const result = await client.call("add", [2, 3]);
    expect(result).toBe(5);
  });

  test("TypeScript HTTP client can call Go HTTP server - echo", async () => {
    const session = new Session();
    const client = new HttpClient(`http://localhost:${HTTP_PORT}`, session);

    const data = { message: "hello", value: 42 };
    const result = await client.call("echo", data);
    expect(result).toEqual(data);
  });

  test("TypeScript HTTP client can use object references from Go server", async () => {
    const session = new Session();
    const client = new HttpClient(`http://localhost:${HTTP_PORT}`, session);

    // Create counter on Go server
    const counter = await client.call("createCounter");
    expect(counter).toHaveProperty("$ref");

    const ref = (counter as { $ref: string }).$ref;

    // Call increment
    const val1 = await client.call("increment", undefined, ref);
    expect(val1).toBe(1);

    const val2 = await client.call("increment", undefined, ref);
    expect(val2).toBe(2);

    // Get value
    const finalValue = await client.call("getValue", undefined, ref);
    expect(finalValue).toBe(2);
  });

  test("TypeScript HTTP client can send notifications to Go server", async () => {
    const session = new Session();
    const client = new HttpClient(`http://localhost:${HTTP_PORT}`, session);

    // Should not throw
    await client.notify("add", [1, 2]);
    expect(true).toBe(true);
  });

  test("TypeScript HTTP client can send batch requests to Go server", async () => {
    const session = new Session();
    const client = new HttpClient(`http://localhost:${HTTP_PORT}`, session);

    const requests = [
      newRequest("add", [1, 1], 1, undefined),
      newRequest("add", [2, 2], 2, undefined),
      newRequest("add", [3, 3], 3, undefined),
    ];

    const responses = await client.batch(requests);
    expect(responses).toHaveLength(3);

    const results = responses.map((r) => r.result);
    expect(results).toEqual([2, 4, 6]);
  });
});

describe("Cross-Language WebSocket", () => {
  test("TypeScript WS client can call Go WS server - basic add", async () => {
    const session = new Session();
    const client = new WsClient(`ws://localhost:${WS_PORT}`, session);
    await client.connect();

    const result = await client.call("add", [5, 7]);
    expect(result).toBe(12);

    client.close();
  });

  test("TypeScript WS client can call Go WS server - echo", async () => {
    const session = new Session();
    const client = new WsClient(`ws://localhost:${WS_PORT}`, session);
    await client.connect();

    const data = { foo: "bar", num: 123 };
    const result = await client.call("echo", data);
    expect(result).toEqual(data);

    client.close();
  });

  test("TypeScript WS client can use object references from Go server", async () => {
    const session = new Session();
    const client = new WsClient(`ws://localhost:${WS_PORT}`, session);
    await client.connect();

    // Create counter on Go server
    const counter = await client.call("createCounter");
    expect(counter).toHaveProperty("$ref");

    const ref = (counter as { $ref: string }).$ref;

    // Call increment multiple times
    const val1 = await client.call("increment", undefined, ref);
    expect(val1).toBe(1);

    const val2 = await client.call("increment", undefined, ref);
    expect(val2).toBe(2);

    const val3 = await client.call("increment", undefined, ref);
    expect(val3).toBe(3);

    client.close();
  });

  test("TypeScript WS client can send concurrent requests to Go server", async () => {
    const session = new Session();
    const client = new WsClient(`ws://localhost:${WS_PORT}`, session);
    await client.connect();

    const promises = [
      client.call("add", [1, 1]),
      client.call("add", [2, 2]),
      client.call("add", [3, 3]),
      client.call("add", [4, 4]),
      client.call("add", [5, 5]),
    ];

    const results = await Promise.all(promises);
    expect(results).toEqual([2, 4, 6, 8, 10]);

    client.close();
  });

  test("TypeScript WS client can send batch requests to Go server", async () => {
    const session = new Session();
    const client = new WsClient(`ws://localhost:${WS_PORT}`, session);
    await client.connect();

    const requests = [
      newRequest("add", [10, 20], 1, undefined),
      newRequest("add", [30, 40], 2, undefined),
    ];

    const responses = await client.batch(requests);
    expect(responses).toHaveLength(2);

    expect(responses[0]!.result).toBe(30);
    expect(responses[1]!.result).toBe(70);

    client.close();
  });
});

describe("Cross-Language Peer", () => {
  test("TypeScript Peer can call Go Peer - basic add", async () => {
    // Spawn Go server in stdio mode
    const proc = Bun.spawn(
      ["../go/jsonrpc3/testserver", "-stdio"],
      {
        stdin: "pipe",
        stdout: "pipe",
        stderr: "pipe",
      }
    );

    // Wrap stdin in a WritableStream
    const writable = new WritableStream<Uint8Array>({
      write(chunk) {
        proc.stdin.write(chunk);
      },
      close() {
        proc.stdin.end();
      },
    });

    const session = new Session();
    const root = new MethodMap();
    const peer = new Peer(proc.stdout, writable, root, { session });

    // Give peer time to initialize
    await new Promise((resolve) => setTimeout(resolve, 100));

    const result = await peer.call("add", [10, 15]);
    expect(result).toBe(25);

    peer.close();
    proc.kill();
  });

  test("TypeScript Peer can call Go Peer - echo", async () => {
    const proc = Bun.spawn(
      ["../go/jsonrpc3/testserver", "-stdio"],
      {
        stdin: "pipe",
        stdout: "pipe",
        stderr: "pipe",
      }
    );

    const writable = new WritableStream<Uint8Array>({
      write(chunk) {
        proc.stdin.write(chunk);
      },
      close() {
        proc.stdin.end();
      },
    });

    const session = new Session();
    const root = new MethodMap();
    const peer = new Peer(proc.stdout, writable, root, { session });

    await new Promise((resolve) => setTimeout(resolve, 100));

    const data = { test: "data", number: 999 };
    const result = await peer.call("echo", data);
    expect(result).toEqual(data);

    peer.close();
    proc.kill();
  });

  test("TypeScript Peer can use object references from Go server", async () => {
    const proc = Bun.spawn(
      ["../go/jsonrpc3/testserver", "-stdio"],
      {
        stdin: "pipe",
        stdout: "pipe",
        stderr: "pipe",
      }
    );

    const writable = new WritableStream<Uint8Array>({
      write(chunk) {
        proc.stdin.write(chunk);
      },
      close() {
        proc.stdin.end();
      },
    });

    const session = new Session();
    const root = new MethodMap();
    const peer = new Peer(proc.stdout, writable, root, { session });

    await new Promise((resolve) => setTimeout(resolve, 100));

    // Create counter on Go server
    const counter = await peer.call("createCounter");
    expect(counter).toHaveProperty("$ref");

    const ref = (counter as { $ref: string }).$ref;

    // Increment
    const val1 = await peer.call("increment", undefined, ref);
    expect(val1).toBe(1);

    const val2 = await peer.call("increment", undefined, ref);
    expect(val2).toBe(2);

    peer.close();
    proc.kill();
  });
});

File: ruby/test/cross_language_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "socket"

# Cross-language integration tests
# Tests Ruby clients against Go servers
class CrossLanguageTest < Minitest::Test
  HTTP_PORT = 18083
  WS_PORT = 18084
  PEER_PORT = 18085
  GO_SERVER_PROCESS = nil

  def self.startup
    # Start Go test server
    go_server_path = File.expand_path("../../go/jsonrpc3/testserver", __dir__)

    unless File.exist?(go_server_path)
      raise "Go test server not found at #{go_server_path}. Run 'go build' in go/jsonrpc3/cmd/testserver"
    end

    @@go_server_process = spawn(
      go_server_path,
      "-http", HTTP_PORT.to_s,
      "-ws", WS_PORT.to_s,
      "-peer", PEER_PORT.to_s,
      out: File::NULL,
      err: File::NULL
    )

    # Wait for HTTP server to be ready
    30.times do
      begin
        socket = TCPSocket.new("localhost", HTTP_PORT)
        socket.close
        break
      rescue Errno::ECONNREFUSED, Errno::EADDRNOTAVAIL
        sleep 0.1
      end
    end

    # Wait for WebSocket server to be ready
    sleep 0.5

    puts "Go test server started on HTTP port #{HTTP_PORT}, WS port #{WS_PORT}"
  end

  def self.shutdown
    if defined?(@@go_server_process) && @@go_server_process
      Process.kill("TERM", @@go_server_process)
      Process.wait(@@go_server_process)
      puts "Go test server stopped"
    end
  end

  def test_ruby_http_client_can_call_go_server_add
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    result = client.call("add", [2, 3])
    assert_equal 5, result
  end

  def test_ruby_http_client_can_call_go_server_echo
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    data = { "message" => "hello", "value" => 42 }
    result = client.call("echo", data)
    assert_equal data, result
  end

  def test_ruby_http_client_can_use_batch_with_references
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    # Use batch to keep everything in same request
    results = JSONRPC3.execute_batch(client) do |b|
      counter = b.call("createCounter", nil)
      counter.call("increment", nil)
      counter.call("increment", nil)
      counter.call("getValue", nil)
    end

    assert_equal 4, results.length

    # First result should be a reference
    counter_ref = results.get_result(0)
    assert counter_ref.is_a?(Hash)
    assert counter_ref.key?("$ref")

    # Subsequent results should be increment values
    assert_equal 1, results.get_result(1)
    assert_equal 2, results.get_result(2)
    assert_equal 2, results.get_result(3)
  end

  def test_ruby_http_client_can_send_notifications
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    # Should not raise
    client.notify("add", [1, 2])
    assert true
  end

  def test_ruby_http_client_can_send_batch_requests
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    requests = [
      JSONRPC3::Request.new(jsonrpc: "3.0", method: "add", params: [1, 1], id: 1),
      JSONRPC3::Request.new(jsonrpc: "3.0", method: "add", params: [2, 2], id: 2),
      JSONRPC3::Request.new(jsonrpc: "3.0", method: "add", params: [3, 3], id: 3)
    ]

    responses = client.batch(requests)
    assert_equal 3, responses.length

    results = responses.map(&:result)
    assert_equal [2, 4, 6], results
  end

  def test_ruby_http_client_can_use_batch_builder
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    results = JSONRPC3.execute_batch(client) do |b|
      b.call("add", [1, 2])
      b.call("add", [3, 4])
      b.call("add", [5, 6])
    end

    assert_equal 3, results.length
    assert_equal 3, results.get_result(0)
    assert_equal 7, results.get_result(1)
    assert_equal 11, results.get_result(2)
  end

  def test_ruby_client_can_query_go_server_methods
    client = JSONRPC3::HttpClient.new("http://localhost:#{HTTP_PORT}")

    # Query available methods
    methods = client.call("$methods")
    assert methods.is_a?(Array)

    # Verify methods is an array of objects
    method_names = methods.map { |m| m["name"] }
    assert_includes method_names, "add"
    assert_includes method_names, "echo"
    assert_includes method_names, "createCounter"
    assert_includes method_names, "$methods"
    assert_includes method_names, "$type"

    # Find add method and check it has metadata
    add_method = methods.find { |m| m["name"] == "add" }
    assert add_method, "add method should be in result"
    assert_equal "Adds a list of numbers", add_method["description"]
    assert_equal ["number"], add_method["params"]
  end

  def test_ruby_client_can_query_go_counter_introspection
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Create a counter
    counter_ref = client.call("createCounter")
    assert counter_ref.is_a?(Hash)
    assert counter_ref.key?("$ref")

    # Query counter's type
    type = client.call("$type", nil, counter_ref)
    assert_equal "Counter", type

    # Query counter's methods
    methods = client.call("$methods", nil, counter_ref)
    method_names = methods.map { |m| m["name"] }
    assert_includes method_names, "increment"
    assert_includes method_names, "getValue"

    # Verify increment method has metadata
    increment_method = methods.find { |m| m["name"] == "increment" }
    assert increment_method, "increment method should be in result"
    assert_equal "Increments the counter by 1", increment_method["description"]

    # Verify getValue method has metadata
    get_value_method = methods.find { |m| m["name"] == "getValue" }
    assert get_value_method, "getValue method should be in result"
    assert_equal "Gets the current counter value", get_value_method["description"]
  end

  # WebSocket Tests

  def test_ruby_ws_client_can_call_go_server_add
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    result = client.call("add", [2, 3])
    assert_equal 5, result

    client.close
  end

  def test_ruby_ws_client_can_call_go_server_echo
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    data = { "message" => "hello", "value" => 42 }
    result = client.call("echo", data)
    assert_equal data, result

    client.close
  end

  def test_ruby_ws_client_can_create_and_use_counter
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Create counter
    counter_ref = client.call("createCounter")
    assert counter_ref.is_a?(Hash)
    assert counter_ref.key?("$ref")

    # Increment
    val1 = client.call("increment", nil, counter_ref)
    assert_equal 1, val1

    val2 = client.call("increment", nil, counter_ref)
    assert_equal 2, val2

    # Get value
    final = client.call("getValue", nil, counter_ref)
    assert_equal 2, final

    client.close
  end

  def test_ruby_ws_client_can_send_notifications
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Should not raise
    client.notify("add", [1, 2])
    assert true

    client.close
  end

  def test_ruby_ws_client_can_query_go_server_methods
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Query available methods
    methods = client.call("$methods")
    assert methods.is_a?(Array)

    # Verify methods is an array of objects
    method_names = methods.map { |m| m["name"] }
    assert_includes method_names, "add"
    assert_includes method_names, "echo"
    assert_includes method_names, "createCounter"
    assert_includes method_names, "$methods"
    assert_includes method_names, "$type"

    # Find add method and check it has metadata
    add_method = methods.find { |m| m["name"] == "add" }
    assert add_method, "add method should be in result"
    assert_equal "Adds a list of numbers", add_method["description"]
    assert_equal ["number"], add_method["params"]

    client.close
  end

  def test_ruby_ws_client_can_query_go_counter_introspection
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Create a counter
    counter_ref = client.call("createCounter")
    assert counter_ref.is_a?(Hash)
    assert counter_ref.key?("$ref")

    # Query counter's type
    type = client.call("$type", nil, counter_ref)
    assert_equal "Counter", type

    # Query counter's methods
    methods = client.call("$methods", nil, counter_ref)
    method_names = methods.map { |m| m["name"] }
    assert_includes method_names, "increment"
    assert_includes method_names, "getValue"

    # Verify increment method has metadata
    increment_method = methods.find { |m| m["name"] == "increment" }
    assert increment_method, "increment method should be in result"
    assert_equal "Increments the counter by 1", increment_method["description"]

    client.close
  end

  def test_ruby_ws_client_concurrent_requests
    client = JSONRPC3::WebSocketClient.new("ws://localhost:#{WS_PORT}")

    # Make multiple concurrent requests
    threads = 5.times.map do |i|
      Thread.new do
        result = client.call("add", [i, i])
        assert_equal i * 2, result
      end
    end

    threads.each(&:join)

    client.close
  end

end

# Run startup/shutdown hooks
Minitest.after_run do
  CrossLanguageTest.shutdown
end

CrossLanguageTest.startup
